scripts/typemock/src/arg.ts

```diff
@@ -0,0 +1,298 @@
+    /**
+     * Allows any value with the provided `typeof` tag.
+     */
+    public static typeof<T = any>(tag: string): T & Arg;
+    public static typeof(tag: string): any {
+        return <any>new Arg(value => typeof value === tag, `typeof ${tag}`);
+    }
+
+    public static string() { return this.typeof("string"); }
+    public static number() { return this.typeof("number"); }
+    public static boolean() { return this.typeof("boolean"); }
+    public static symbol() { return this.typeof("symbol"); }
+    public static object() { return this.typeof("object"); }
+    public static function() { return this.typeof("function"); }
+
+    /**
+     * Allows any value that is an instance of the provided function.
+     * @param type The expected constructor.
+     */
+    public static instanceof<TClass extends { new (...args: any[]): object; prototype: object; }>(type: TClass): TClass["prototype"] & Arg {
+        return <any>new Arg(value => value instanceof type, `instanceof ${type.name}`);
+    }
+
+    /**
+     * Allows any value that has the provided property names in its prototype chain.
+     */
+    public static has<T>(...names: string[]): T & Arg {
+        return <any>new Arg(value => names.filter(name => name in value).length === names.length, `has ${names.join(", ")}`);
+    }
+
+    /**
+     * Allows any value that has the provided property names on itself but not its prototype chain.
+     */
+    public static hasOwn<T>(...names: string[]): T & Arg {
+        return <any>new Arg(value => names.filter(name => Object.prototype.hasOwnProperty.call(value, name)).length === names.length, `hasOwn ${names.join(", ")}`);
+    }
+
+    /**
+     * Allows any value that matches the provided condition for the rest of the arguments in the call.
+     * @param condition The optional condition for each other element.
+     */
+    public static rest<T>(condition?: T | (T & Arg)): T & Arg {
+        if (condition === undefined) {
+            return <any>new Arg(() => true, `rest`, /*rest*/ true);
+        }
+
+        const arg = Arg.from(condition);
+        return <any>new Arg(value => arg._validate(value), `rest ${arg._message}`, /*rest*/ true);
+    }
+
+    /**
+     * Negates a condition.
+     */
+    public static not<T = any>(value: T | (T & Arg)): T & Arg {
+        const arg = Arg.from(value);
+        return <any>new Arg(value => !arg._validate(value), `not ${arg._message}`);
+    }
+
+    /**
+     * Combines conditions, where all conditions must be `true`.
+     */
+    public static and<T = any>(...args: ((T & Arg) | T)[]): T & Arg {
+        const conditions = args.map(Arg.from);
+        return <any>new Arg(value => conditions.every(condition => condition._validate(value)), conditions.map(condition => condition._message).join(" and "));
+    }
+
+    /**
+     * Combines conditions, where no condition may be `true`.
+     */
+    public static nand<T = any>(...args: ((T & Arg) | T)[]): T & Arg {

```

Comment:
Philosophical nit: Why include `nand` and `nor` primitives? Most code is just written using `and`, `or`, and `not`, because the compound operators are harder to reason about the truth table of. In any case, if you're going for including all the bells, I feel like implication is a missing truth table that's as likely to provide as much value as `nand`, at least.